feat(CxMembership): Set Catena-X Membership in BPDM

tfjanjua opened the pull request:
## Description

Set Catena-X Membership in BPDM.

## Why

As Portal, we would like to maintain a list of all companies that were onboarded to the Catena-X network. This membership information is meant to be stored in the BPDM Pool. We therefore need to update the `isCatenaXMember` flag for a BPNL once the company has been onboarded.

## Issue

Ref: #1111

Config PR: https://github.com/eclipse-tractusx/portal/pull/472

BPDM PR: https://github.com/eclipse-tractusx/bpdm/pull/1072

## Checklist

- [x] I have followed the [contributing guidelines](https://github.com/eclipse-tractusx/portal-assets/blob/main/docs/developer/Technical%20Documentation/Dev%20Process/How%20to%20contribute.md#commit-and-pr-guidelines)
- [x] I have performed a self-review of my own code
- [x] I have successfully tested my changes locally
- [x] I have added tests that prove my changes work
- [x] I have checked that new and existing tests pass locally with my changes
- [x] I have commented my code, particularly in hard-to-understand areas

ntruchsess commented on src/externalsystems/Bpdm.Library/BpdmService.cs:
```diff
@@ -176,4 +177,23 @@ public async Task<BpdmSharingState> GetSharingState(Guid applicationId, Cancella
+
+        // Same User (which we have for BPDM service) can be used to call Business Partner Pool
+        // But BaseAddress of Business Partner Pool is different as its deployed on another server.
+        httpClient.BaseAddress = new Uri(_settings.BusinessPartnerPoolBaseAddress);
```
(while technically working) this is not how a httpClientFactory (which is used in tokenService) should be used. The BusinesCode should not have to deal with the baseAddress.

tfjanjua replied:
@ntruchsess I totally agree to not assign value to `BaseAddress` param at library when its already managed in common token service but here we have a situation that we want to call this newly introduced method `SetCxMembership` from background process and not from the controller (because there we have a possibility to take user's token use same token to call Business Partner Pool api route) and same user (which has been used to call BPDM services) can be used to call Business Partner Pool service so, thats why I introduced another kind of baseAddress `BusinessPartnerPoolBaseAddress` within the same BPDM configs and override the value of BaseAddress of BPDM config before calling the tokenService so, I could use same user (which has been used to call BPDM services) and could get the token to call this Business Partner Pool service.

It would be nice if we could separate out BpnAccess service from BPDM library because I think then we can avoid assigning or overriding the value of BaseAddress because then we would have opportunity to have proper user (with baseAddress) to call Business Partner Pool services (no matter from controller or background process).

Phil91 replied:
@tfjanjua I agree with @ntruchsess instead of changing the url in the business logic we should adjust the `BpnAccess` from using the token of the user to using a technical user. Therefore we would need to add a technical user with the needed roles (we should discuss with @nicoprow which roles would be needed to set it up properly) after that we can use the bpnAccess for the web services as well as the process worker.

ntruchsess reviewed (changes requested):
I did fix a few findings and force-pushed to your repo after rebasing on latest main. Please pull he changes before proceeding. (And also have a look on what I changed to avoid such findings in the future).
PR currently cannot be merged as the configuration in portal-repo does not match. Also we need to come up with a better approach to handle multiple bpdm deployments without hardcoding the fact that this particular call is different into the service-method.

tfjanjua commented:
> > @tfjanjua : I did fix a few findings and force-pushed to your repo after rebasing on latest main. Please pull he changes before proceeding. (And also have a look on what I changed to avoid such findings in the future). PR currently cannot be merged as the configuration in portal-repo does not match. Also we need to come up with a better approach to handle multiple bpdm deployments without hardcoding the fact that this particular call is different into the service-method.
> 
> Thank you @ntruchsess , providing your feedback.
> 
> I've looked at your commit and found some formatting and addition of license related information (especially in migration files).
> 
> I'll have a look at configuration PR.

@ntruchsess , I have added my comment in the corresponding config PR: https://github.com/eclipse-tractusx/portal/pull/472#discussion_r1854239204

You may resolve the conversation if the provided explanations make sense to you as well.

Thanks

evegufy commented:
@tfjanjua could you please resolve the conflict?
@ntruchsess @Phil91 could you please review?

tfjanjua commented:
> @tfjanjua : I did fix a few findings and force-pushed to your repo after rebasing on latest main. Please pull he changes before proceeding. (And also have a look on what I changed to avoid such findings in the future). PR currently cannot be merged as the configuration in portal-repo does not match. Also we need to come up with a better approach to handle multiple bpdm deployments without hardcoding the fact that this particular call is different into the service-method.

Hi @ntruchsess , I forgot to take pull before resolving the conflicts and lost your finding's fix but I have tried to find your findings (like removing bom from utf-8 encoding and adding a proper license info especially on newly created files) and tried to fix them, please let me know if I have missed any point.

tfjanjua commented:
@evegufy | @Phil91 | @ntruchsess 
Please let me know if there is any other finding so, we can resolve them and merge this PR as soon as possible. Thanks

evegufy commented:
@tfjanjua could you please solve the conflict?
@Phil91 @ntruchsess could you please review? The PR is open since quite some time now.

ntruchsess commented:
@Phil91 : regarding your comment https://github.com/eclipse-tractusx/portal-backend/pull/1118#discussion_r1930683096
from what I can see the new code is not using the current user's token - it's using whatever is configured as technical user in the BpdmServiceSettings,

ntruchsess reviewed (approved):
@Phil91 : regarding your comment https://github.com/eclipse-tractusx/portal-backend/pull/1118#discussion_r1930683096
from what I can see the new code is not using the current user's token - it's using whatever is configured as technical user in the BpdmServiceSettings, So from my point of view this part is fine.
@tfjanjua + @Phil91 : I did refactor the TokenService to support more httpclient-names than just the classname per set of technical-user-credentials. This allows to set the baseUrl for bpdm-pool-service in the httpClient-configuration (removing is from the businessLogic). To solve the conflicts I did also rebase on main and recreated the migration doing a force-push to your repository. Please pull those changes before doing any further changes. (From my point of view no changes should be required though)
